src/core/server/server.ts

```diff
@@ -16,37 +16,56 @@
  * specific language governing permissions and limitations
  * under the License.
  */
+import { Observable } from 'rxjs';
+import { Type } from '@kbn/config-schema';
 
-import { ConfigService, Env } from './config';
+import { ConfigService, Env, Config, ConfigPath } from './config';
 import { ElasticsearchService } from './elasticsearch';
 import { HttpService, HttpServiceSetup, Router } from './http';
 import { LegacyService } from './legacy';
 import { Logger, LoggerFactory } from './logging';
-import { PluginsService } from './plugins';
+import { PluginsService, config as pluginsConfig } from './plugins';
+
+import { config as elasticsearchConfig } from './elasticsearch';
+import { config as httpConfig } from './http';
+import { config as loggingConfig } from './logging';
+import { config as devConfig } from './dev';
 
 export class Server {
+  public readonly configService: ConfigService;
   private readonly elasticsearch: ElasticsearchService;
   private readonly http: HttpService;
   private readonly plugins: PluginsService;
   private readonly legacy: LegacyService;
   private readonly log: Logger;
 
-  constructor(configService: ConfigService, logger: LoggerFactory, env: Env) {
-    const core = { env, configService, logger };
-    this.log = logger.get('server');
+  constructor(
+    readonly config$: Observable<Config>,
+    readonly env: Env,
+    private readonly logger: LoggerFactory
+  ) {
+    this.log = this.logger.get('server');
+    this.configService = new ConfigService(config$, env, logger);
+
+    const core = { configService: this.configService, env, logger };
     this.http = new HttpService(core);
     this.plugins = new PluginsService(core);
     this.legacy = new LegacyService(core);
     this.elasticsearch = new ElasticsearchService(core);
   }
 
+  public async preSetup() {
```

Comment:
__suggestion:__ honestly I'd just make `setupConfigSchemas` public and get rid of `preSetup` - it's clearer what it does and there is no strong evidence yet that we'll want to have a generic `preSetup` hook in core. But I'll defer that decision to you.